auth/auth.go

```diff
@@ -309,39 +309,62 @@ func (auth *Authenticator) UpdateSequenceNumber(p Principal, seq uint64) error {
 }
 
 // Invalidates the channel list of a user/role by saving its Channels() property as nil.
-func (auth *Authenticator) InvalidateChannels(p Principal, invalSeq uint64) error {
-	invalidateChannelsCallback := func(p Principal) (updatedPrincipal Principal, err error) {
+func (auth *Authenticator) InvalidateChannels(name string, isUser bool, invalSeq uint64) error {
+	for i := 1; i < PrincipalUpdateMaxCasRetries; i++ {
+		var princ Principal
+		var docID string
+
+		if isUser {
+			princ = &userImpl{}
+			docID = docIDForUser(name)
+		} else {
+			princ = &roleImpl{}
+			docID = docIDForRole(name)
+		}
 
-		if p == nil || p.Channels() == nil {
-			return p, base.ErrUpdateCancel
+		cas, err := auth.bucket.Get(docID, &princ)
+		if err != nil {
+			return err
 		}
 
-		base.Infof(base.KeyAccess, "Invalidate access of %q", base.UD(p.Name()))
+		princ.SetChannelInvalSeq(invalSeq)
```

Comment:
Can you remind me why we need to set invalSeq to the most recent invalidation, instead of the earliest invalidation, in the case where there are multiple invalidations without an auth?

Reply:
We need to use the earliest. I missed it here, thanks for catching!